pkg/deploy/controller/deployment/controller_test.go

```diff
@@ -651,6 +651,73 @@ func TestHandle_deployerPodDisappeared(t *testing.T) {
 	}
 }
 
+func expectMapContains(t *testing.T, exists, expected map[string]string, testCase, what string) {
+	if expected == nil {
+		return
+	}
+	for k, v := range expected {
+		value, ok := exists[k]
+		if ok && value != v {
+			t.Errorf("%s expected %s[%s]=%s, got %s", testCase, what, k, v, value)
+		} else if !ok {
+			t.Errorf("%s expected %s %s: not present", testCase, what, k)
+		}
+	}
+}
+
+func TestDeployerCustomLabelsAndAnnotations(t *testing.T) {
+	controller := &DeploymentController{
+		decodeConfig: func(deployment *kapi.ReplicationController) (*deployapi.DeploymentConfig, error) {
+			return deployutil.DecodeDeploymentConfig(deployment, api.Codec)
+		},
+		podClient: &podClientImpl{
+			createPodFunc: func(namespace string, pod *kapi.Pod) (*kapi.Pod, error) {
+				return pod, nil
+			},
+		},
+		makeContainer: func(strategy *deployapi.DeploymentStrategy) (*kapi.Container, error) {
+			return okContainer(), nil
+		},
+	}
+
+	testCases := []struct {
+		name         string
+		strategy     deployapi.DeploymentStrategy
+		labels       map[string]string
+		annotations  map[string]string
+		verifyLabels bool
+	}{
+		{name: "labels and annotations", strategy: deploytest.OkStrategy(), labels: map[string]string{"label1": "value1"}, annotations: map[string]string{"annotation1": "value1"}, verifyLabels: true},
+		{name: "custom strategy, no annotations", strategy: deploytest.OkCustomStrategy(), labels: map[string]string{"label2": "value2", "label3": "value3"}, verifyLabels: true},
+		{name: "custom strategy, no labels", strategy: deploytest.OkCustomStrategy(), annotations: map[string]string{"annotation3": "value3"}, verifyLabels: true},
+		{name: "no overrride", strategy: deploytest.OkStrategy(), labels: map[string]string{deployapi.DeployerPodForDeploymentLabel: "ignored"}, verifyLabels: false},
+	}
+
+	for _, test := range testCases {
```

Comment:
Not required, but for future reference it's simpler to just do something like `t.Logf("evaluating test %q", test.name)` instead of passing the name around everywhere since the tests are sequential (errors for a test will appear right under that test's name)